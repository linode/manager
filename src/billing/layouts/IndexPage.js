import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';

import { setSource } from '~/actions/source';
import { setTitle } from '~/actions/title';

import { Tabs } from 'linode-components/tabs';


export class IndexPage extends Component {
  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));

    dispatch(setTitle('Billing'));
  }

  render() {
    const { dispatch, children } = this.props;

    const tabs = [
      { name: 'Dashboard', link: '' },
    ].map(function (tab) {
      return {
        ...tab,
        link: `/billing${tab.link}`,
      };
    });

    return (
      <div>
        <header className="main-header">
          <div className="container">
            <h1>Billing</h1>
          </div>
        </header>
        <div className="main-header-fix"></div>
        <Tabs
          tabs={tabs}
          onClick={(e, tabIndex) => {
            e.stopPropagation();
            dispatch(push(tabs[tabIndex].link));
          }}
          pathname={location.pathname}
        >{children}</Tabs>
      </div>
    );
  }
}

IndexPage.propTypes = {
  children: PropTypes.node.isRequired,
  dispatch: PropTypes.func,
};

function mapStateToProps() {
  return {};
}

export default connect(mapStateToProps)(IndexPage);
