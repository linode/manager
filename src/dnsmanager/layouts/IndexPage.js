import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';

import { setError } from '~/actions/errors';
import { showModal, hideModal } from '~/actions/modal';
import ConfirmModalBody from '~/components/modals/ConfirmModalBody';
import { dnszones } from '~/api';
import { setSource } from '~/actions/source';
import { setTitle } from '~/actions/title';
import CreateHelper from '~/components/CreateHelper';
import Checkbox from '~/components/Checkbox';

export class IndexPage extends Component {
  constructor() {
    super();
    this.deleteZone = this.deleteZone.bind(this);
  }

  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));
    dispatch(setTitle('DNS Manager'));
    try {
      await dispatch(dnszones.all());
    } catch (response) {
      dispatch(setError(response));
    }
  }

  deleteZone(zoneId) {
    const { dispatch } = this.props;
    dispatch(showModal('Delete DNS Zone', this.renderModal(zoneId)));
  }

  renderModal(zoneId) {
    const { dispatch } = this.props;
    return (
      <ConfirmModalBody
        buttonText="Delete"
        onOk={async () => {
          await dispatch(dnszones.delete(zoneId));
          dispatch(hideModal());
        }}
        onCancel={() => dispatch(hideModal())}
      >
        <span className="text-danger">WARNING!</span> This will permanently
        delete this DNS Zone. Confirm below to proceed.
      </ConfirmModalBody>
    );
  }

  render() {
    const { dnszones } = this.props;

    const renderZones = () => (
      <section>
        <table>
          <thead>
            <tr>
              <th>&nbsp;</th>
              <th>Name</th>
              <th>Type</th>
              <th>&nbsp;</th>
            </tr>
          </thead>
          <tbody>
            {Object.values(dnszones.dnszones).map(z => (
              <tr key={z.id}>
                <td>
                  <Checkbox />
                </td>
                <td>
                  <Link to={`/dnsmanager/${z.id}`}>{z.dnszone}</Link>
                </td>
                <td>{z.type}</td>
                <td className="text-xs-right">
                  <a href="#" onClick={() => this.deleteZone(z.id)}>Delete</a>
                </td>
              </tr>
             ))}
          </tbody>
        </table>
      </section>
    );

    return (
      <div className="container">
        <header className="clearfix">
          <div className="float-xs-left">
            <h1>DNS Manager</h1>
          </div>
          <div className="float-xs-right">
            <Link to="/dnsmanager/create" className="btn btn-default">Add a zone</Link>
          </div>
        </header>
        {Object.keys(this.props.dnszones.dnszones).length ? renderZones() :
          <CreateHelper label="zones" href="/dnsmanager/create" linkText="Add a zone" />}
      </div>
    );
  }
}

IndexPage.propTypes = {
  dispatch: PropTypes.func,
  dnszones: PropTypes.object,
  selected: PropTypes.object,
};


function select(state) {
  return {
    dnszones: state.api.dnszones,
  };
}

export default connect(select)(IndexPage);
