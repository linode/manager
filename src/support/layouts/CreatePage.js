import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import { push } from 'react-router-redux';

import { Card } from 'linode-components/cards';
import {
  Input,
  Select,
  Form,
  FormGroup,
  FormGroupError,
  SubmitButton,
} from 'linode-components/forms';

import { setSource } from '~/actions/source';
import { setTitle } from '~/actions/title';
import { domains, linodes, nodebalancers, tickets } from '~/api';
import { dispatchOrStoreErrors, FormSummary } from '~/components/forms';

import TicketHelper from '../components/TicketHelper';


export class CreatePage extends Component {
  static async preload({ dispatch }) {
    await Promise.all([
      dispatch(linodes.all()),
      dispatch(domains.all()),
      dispatch(nodebalancers.all()),
    ]);
  }

  constructor(props) {
    super(props);

    this.state = {
      summary: '',
      regarding: `linode_id:${Object.values(props.linodes)[0].id}`,
      description: '',
      errors: {},
      loading: false,
    };
  }

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));

    dispatch(setTitle('Open a Ticket'));
  }

  onSubmit = () => {
    const { dispatch } = this.props;
    const { summary, regarding, description } = this.state;

    const regardingField = regarding.substring(0, regarding.indexOf(':'));
    const regardingId = regarding.substring(regardingField.length + 1);

    this.setState({ loading: true, errors: {} });

    return dispatch(dispatchOrStoreErrors.call(this, [
      () => tickets.post({ summary, description, [regardingField]: +regardingId }),
      ({ id }) => push(`/support/${id}`),
    ]));
  }

  onChange = ({ target: { name, value } }) => this.setState({ [name]: value })

  renderOptionsGroup(label, field, group) {
    return (
      <optgroup label={label} key={label}>
        {group.map(object => (
          <option key={object.id} value={`${field}:${object.id}`}>
            {object.label || object.domain}
          </option>
         ))}
      </optgroup>
    );
  }

  render() {
    const { summary, regarding, description, loading, errors } = this.state;
    const { linodes, domains, nodebalancers } = this.props;
    const regardingOptions = [
      this.renderOptionsGroup('Linodes', 'linode_id', Object.values(linodes)),
      this.renderOptionsGroup('Domains', 'domain_id', Object.values(domains)),
      this.renderOptionsGroup('NodeBalancers', 'nodebalancer_id', Object.values(nodebalancers)),
      // TODO: this is not currently supported by the API
      // this.renderOptionsGroup('Other', [{ label: 'Other', id: 'other' }]),
    ];

    return (
      <div className="container create-page">
        <header className="text-sm-left">
          <Link to="/support">Support</Link>
          <h1>Open a ticket</h1>
        </header>
        <section>
          <TicketHelper displayHeader />
        </section>
        <Card>
          <Form onSubmit={this.onSubmit}>
            <FormGroup className="row" errors={errors} name="summary">
              <label htmlFor="summary" className="col-sm-2 col-form-label">Summary</label>
              <div className="col-sm-10">
                <Input
                  name="summary"
                  id="summary"
                  value={summary}
                  className="input-lg"
                  onChange={this.onChange}
                />
                <FormGroupError errors={errors} name="summary" inline={false} />
              </div>
            </FormGroup>
            <FormGroup className="row" errors={errors} name="regarding">
              <label htmlFor="regarding" className="col-sm-2 col-form-label">Regarding</label>
              <div className="col-sm-10">
                <Select
                  name="regarding"
                  id="regarding"
                  value={regarding}
                  onChange={this.onChange}
                >{regardingOptions}</Select>
                <FormGroupError errors={errors} name="regarding" inline={false} />
              </div>
            </FormGroup>
            <FormGroup className="row" errors={errors} name="description">
              <label htmlFor="description" className="col-sm-2 col-form-label">
                Description
              </label>
              <div className="col-sm-10">
                <textarea
                  className="textarea-md"
                  name="description"
                  id="description"
                  value={description}
                  onChange={this.onChange}
                />
                <FormGroupError errors={errors} name="description" inline={false} />
              </div>
            </FormGroup>
            <FormGroup className="row">
              <div className="col-sm-10 offset-sm-2">
                <SubmitButton disabled={loading} disabledChildren="Opening Ticket">
                  Open Ticket
                </SubmitButton>
                <FormSummary errors={errors} />
              </div>
            </FormGroup>
          </Form>
        </Card>
      </div>
    );
  }
}

CreatePage.propTypes = {
  dispatch: PropTypes.func.isRequired,
  linodes: PropTypes.object.isRequired,
  domains: PropTypes.object.isRequired,
  nodebalancers: PropTypes.object.isRequired,
};

function select(state) {
  return {
    nodebalancers: state.api.nodebalancers.nodebalancers,
    linodes: state.api.linodes.linodes,
    domains: state.api.domains.domains,
  };
}

export default connect(select)(CreatePage);
