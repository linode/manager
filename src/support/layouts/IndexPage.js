import _ from 'lodash';
import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';

import { ListBody, ListGroup } from 'linode-components/lists/bodies';
import { TableCell } from 'linode-components/tables/cells';
import { List } from 'linode-components/lists';
import { Table } from 'linode-components/tables';

import { setSource } from '~/actions/source';
import { setTitle } from '~/actions/title';
import { tickets } from '~/api';
import CreateHelper from '~/components/CreateHelper';
import {
  getLinodeRedirectUrl, getNodebalancerRedirectUrl, getDomainRedirectUrl,
} from '~/components/notifications/EventTypes';
import TimeDisplay from '~/components/TimeDisplay';


const TICKET_LINK_MAP = {
  linode: getLinodeRedirectUrl,
  nodebalancer: getNodebalancerRedirectUrl,
  domain: getDomainRedirectUrl,
};

export function renderTicketCreationInfo(ticket) {
  const entity = ticket.entity || {};
  const to = TICKET_LINK_MAP[entity.type];

  let link = <strong>{entity.label}</strong>;
  if (to) {
    link = <Link to={to(entity)}>{entity.label}</Link>;
  }

  return (
    <div>
      Opened
      {!ticket.opened_by ? null : (
        <span>
          &nbsp;by <strong id={`opened-by-${ticket.id}`}>{ticket.opened_by}</strong>
        </span>
      )}
      &nbsp;<span id={`opened-${ticket.id}`}><TimeDisplay time={ticket.opened} /></span>
      {ticket.entity ? (
        <span>
          &nbsp;regarding <span id={`regarding-${ticket.id}`}>{link}</span>
        </span>
      ) : null}
    </div>
  );
}

export class IndexPage extends Component {
  static async preload({ dispatch }) {
    await dispatch(tickets.all());
  }

  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));

    dispatch(setTitle('Support'));
  }

  renderLabelCell = ({ record: ticket }) => {
    return (
      <TableCell
        column={{ className: 'LabelCell', headerClassName: 'TicketLabelColumn' }}
        record={ticket}
      >
        <Link title={ticket.id} to={`/support/${ticket.id}`}>{ticket.summary}</Link>
        <small>{renderTicketCreationInfo(ticket)}</small>
      </TableCell>
    );
  }

  renderUpdatedByCell({ record: ticket }) {
    return (
      <TableCell column={{}} record={ticket}>
        {!ticket.updated_by ? <span>No updates</span> : (
          <span>
            Updated by <strong id={`updated-by-${ticket.id}`}>{ticket.updated_by}</strong>
            &nbsp;<span id={`updated-${ticket.id}`}><TimeDisplay time={ticket.updated} /></span>
          </span>
        )}
      </TableCell>
    );
  }

  renderTickets(tickets) {
    const statusFormat = {
      new: 'Open',
      open: 'Open',
      closed: 'Closed',
    };

    const groups = _.map(
      _.groupBy(
        _.sortBy(Object.values(tickets), ticket => new Date(ticket.created)).reverse(),
        ({ status }) => statusFormat[status]),
      (_tickets, _group) => ({
        name: _group,
        columns: [
          {
            cellComponent: this.renderLabelCell,
            headerClassName: 'TicketLabelColumn',
          },
          {
            dataKey: 'id',
            className: 'hidden-md-down',
            headerClassName: 'TicketIdColumn hidden-md-down',
            formatFn: id => <span>Ticket #{id}</span>,
          },
          {
            cellComponent: this.renderUpdatedByCell,
          },
        ],
        data: _tickets,
      }));

    if (groups[0].name !== 'Open') {
      groups.unshift({ name: 'Open', data: [] });
    }

    return (
      <List>
        <ListBody>
          {groups.map((group, index) => {
            return (
              <ListGroup
                key={index}
                name={group.name}
              >
                <Table
                  className="Tickets"
                  columns={group.columns}
                  data={group.data}
                  selectedMap={{}}
                  noDataMessage="You have no open tickets."
                  disableHeader
                />
              </ListGroup>
            );
          })}
        </ListBody>
      </List>
    );
  }

  render() {
    const { tickets } = this.props;

    return (
      <div className="PrimaryPage container">
        <header className="PrimaryPage-header">
          <div className="PrimaryPage-headerRow clearfix">
            <h1 className="float-sm-left">Support</h1>
            <Link to="/support/create" className="linode-add btn btn-primary float-sm-right">
              <span className="fa fa-plus"></span>
              Open a ticket
            </Link>
          </div>
        </header>
        <div className="PrimaryPage-body">
          {Object.keys(tickets.tickets).length ? this.renderTickets(tickets.tickets) :
            <CreateHelper label="tickets" href="/support/create" linkText="Open a ticket" />}
        </div>
      </div>
    );
  }
}

IndexPage.propTypes = {
  dispatch: PropTypes.func.isRequired,
  tickets: PropTypes.object.isRequired,
};


function select(state) {
  return {
    tickets: state.api.tickets,
  };
}

export default connect(select)(IndexPage);
