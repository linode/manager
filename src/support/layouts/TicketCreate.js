import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { push } from 'react-router-redux';
import { compose } from 'redux';
import Card from 'linode-components/dist/cards/Card';
import Form from 'linode-components/dist/forms/Form';
import FormGroup from 'linode-components/dist/forms/FormGroup';
import FormGroupError from 'linode-components/dist/forms/FormGroupError';
import FormSummary from 'linode-components/dist/forms/FormSummary';
import Input from 'linode-components/dist/forms/Input';
import Textarea from 'linode-components/dist/forms/Textarea';
import Select from 'linode-components/dist/forms/Select';
import SubmitButton from 'linode-components/dist/forms/SubmitButton';
import { onChange } from 'linode-components/dist/forms/utilities';

import { setAnalytics, setSource } from '~/actions';
import api from '~/api';
import { dispatchOrStoreErrors } from '~/api/util';
import ChainedDocumentTitle from '~/components/ChainedDocumentTitle';
import { ComponentPreload as Preload } from '~/decorators/Preload';

import TicketHelper from '../components/TicketHelper';


export class TicketCreate extends Component {
  constructor(props) {
    super(props);

    this.state = {
      summary: '',
      regarding: null,
      description: '',
      errors: {},
      loading: false,
    };

    this.onChange = onChange.bind(this);
  }

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));
    dispatch(setAnalytics(['tickets', 'create']));
  }

  onSubmit = () => {
    const { dispatch } = this.props;
    const { summary, regarding, description } = this.state;

    const regardingField = regarding.substring(0, regarding.indexOf(':'));
    const regardingId = regarding.substring(regardingField.length + 1);

    const data = { summary, description };

    if (regardingField !== 'other') {
      data[regardingField] = +regardingId;
    }

    return dispatch(dispatchOrStoreErrors.call(this, [
      () => api.tickets.post(data),
      ({ id }) => push(`/support/${id}`),
    ]));
  }

  renderOptionsGroup(label, field, group) {
    return {
      label,
      options: group.map(object => ({
        value: `${field}:${object.id}`,
        label: object.label || object.domain,
      })),
    };
  }

  render() {
    const { summary, regarding, description, loading, errors } = this.state;
    const { linodes, domains, nodebalancers, volumes } = this.props;
    const regardingOptions = [
      this.renderOptionsGroup('Linodes', 'linode_id', Object.values(linodes)),
      this.renderOptionsGroup('Domains', 'domain_id', Object.values(domains)),
      this.renderOptionsGroup('NodeBalancers', 'nodebalancer_id', Object.values(nodebalancers)),
      this.renderOptionsGroup('Volumes', 'volume_id', Object.values(volumes)),
      this.renderOptionsGroup('Other', '', [{ label: 'Other', id: 'other' }]),
    ];

    return (
      <div className="container create-page">
        <ChainedDocumentTitle title="Open a Ticket" />
        <header className="text-sm-left">
          <Link to="/support">Support</Link>
          <h1>Open a ticket</h1>
        </header>
        <section>
          <TicketHelper displayHeader />
        </section>
        <Card>
          <Form
            onSubmit={this.onSubmit}
            analytics={{ title: 'Open a Ticket', action: 'add' }}
          >
            <FormGroup className="row" errors={errors} name="summary">
              <label htmlFor="summary" className="col-sm-2 col-form-label">Summary</label>
              <div className="col-sm-10">
                <Input
                  name="summary"
                  id="summary"
                  value={summary}
                  className="input-lg"
                  onChange={this.onChange}
                />
                <FormGroupError errors={errors} name="summary" inline={false} />
              </div>
            </FormGroup>
            <FormGroup className="row" errors={errors} name="regarding">
              <label htmlFor="regarding" className="col-sm-2 col-form-label">Regarding</label>
              <div className="col-sm-10">
                <Select
                  name="regarding"
                  id="regarding"
                  value={regarding}
                  onChange={this.onChange}
                  options={regardingOptions}
                />
                <FormGroupError errors={errors} name="regarding" inline={false} />
              </div>
            </FormGroup>
            <FormGroup className="row" errors={errors} name="description">
              <label htmlFor="description" className="col-sm-2 col-form-label">
                Description
              </label>
              <div className="col-sm-10">
                <Textarea
                  className="textarea-md"
                  name="description"
                  id="description"
                  value={description}
                  onChange={this.onChange}
                />
                <FormGroupError errors={errors} name="description" inline={false} />
              </div>
            </FormGroup>
            <FormGroup className="row" name="submit">
              <div className="col-sm-10 offset-sm-2">
                <SubmitButton disabled={loading} disabledChildren="Opening Ticket">
                  Open Ticket
                </SubmitButton>
                <FormSummary errors={errors} />
              </div>
            </FormGroup>
          </Form>
        </Card>
      </div>
    );
  }
}

TicketCreate.propTypes = {
  dispatch: PropTypes.func.isRequired,
  linodes: PropTypes.object.isRequired,
  domains: PropTypes.object.isRequired,
  volumes: PropTypes.array.isRequired,
  nodebalancers: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => {
  return {
    nodebalancers: state.api.nodebalancers.nodebalancers,
    linodes: state.api.linodes.linodes,
    domains: state.api.domains.domains,
    volumes: state.api.volumes.volumes,
  };
};

const preloadRequest = async (dispatch) => {
  await Promise.all([
    api.linodes, api.domains, api.nodebalancers, api.volumes,
  ].map(o => dispatch(o.all())));
};

export default compose(
  connect(mapStateToProps),
  Preload(preloadRequest)
)(TicketCreate);
