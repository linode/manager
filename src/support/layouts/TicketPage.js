import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';

import { Card } from 'linode-components/cards';
import {
  Form,
  FormGroup,
  FormGroupError,
  FormSummary,
  SubmitButton,
  Input,
} from 'linode-components/forms';

import { tickets } from '~/api';
import { dispatchOrStoreErrors, getObjectByLabelLazily } from '~/api/util';
import { addTicketAttachment } from '~/api/tickets';
import { setSource } from '~/actions/source';
import { setTitle } from '~/actions/title';
import { MAX_UPLOAD_SIZE_MB } from '~/constants';

import { renderTicketCreationInfo } from './IndexPage';
import TicketReply from '../components/TicketReply';
import TicketHelper from '../components/TicketHelper';


export function AttachmentTooBigError() {
  const error = `File size must be under ${MAX_UPLOAD_SIZE_MB} MB`;
  this.json = () => ({
    errors: [{ field: 'attachments', reason: error }],
  });
}

AttachmentTooBigError.prototype = new Error();

export class TicketPage extends Component {
  static async preload({ dispatch }, { ticketId }) {
    await dispatch(getObjectByLabelLazily('tickets', ticketId, 'id'));
    await dispatch(tickets.replies.all([ticketId]));
  }

  constructor() {
    super();

    this.state = { reply: '', attachments: [], errors: {}, loading: false };
  }

  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));

    dispatch(setTitle(this.props.ticket.summary));
  }

  onChange = ({ target: { name, value } }) => this.setState({ [name]: value })

  onSubmit = () => {
    const { attachments, reply: description } = this.state;
    const { ticket, dispatch } = this.props;

    const requests = [
      // All other requests will get unshift()ed before this so this happens last.
      () => this.setState({ reply: '', attachments: [] }),
    ];

    if (description) {
      requests.unshift(() => tickets.replies.post({ description }, [ticket.id]));
    }

    for (let i = 0; i < attachments.length; i++) {
      const attachment = attachments[i];

      requests.unshift(() => {
        if ((attachment.size / (1024 * 1024)) < MAX_UPLOAD_SIZE_MB) {
          return addTicketAttachment(ticket.id, attachment);
        }

        throw new AttachmentTooBigError();
      });
    }

    return dispatch(dispatchOrStoreErrors.call(this, requests));
  }

  renderTicketResponseForm() {
    const { errors, loading, reply } = this.state;

    return (
      <Card>
        <Form onSubmit={this.onSubmit}>
          <FormGroup>
            <label htmlFor="reply" className="row-label">Write a reply:</label>
            <textarea
              name="reply"
              id="reply"
              className="textarea-lg"
              onChange={this.onChange}
              value={reply}
            />
            <FormGroupError inline={false} errors={errors} name="description" />
          </FormGroup>
          <FormGroup errors={errors}>
            <div>
              <label htmlFor="attachments" className="row-label">Add attachments:</label>
            </div>
            <Input
              multiple
              name="attachments"
              id="attachments"
              type="file"
              onChange={(e) => this.setState({ attachments: e.target.files })}
            />
            <FormGroupError inline={false} errors={errors} name="attachments" />
          </FormGroup>
          <div className="clearfix">
            <div className="text-sm-right">
              <SubmitButton disabled={loading} disabledChildren="Submitting">Submit</SubmitButton>
              <FormSummary errors={errors} success="Response submitted." />
            </div>
          </div>
        </Form>
      </Card>
    );
  }

  renderTicketClosed() {
    return (
      <Card id="ticket-closed">
        This ticket has been closed. If you are still experiencing an issue,
        please <Link to="/support/create">open a new ticket</Link>.
      </Card>
    );
  }

  render() {
    const { ticket, replies } = this.props;
    if (!ticket) {
      return null;
    }

    return (
      <div>
        <header className="main-header main-header--border">
          <div className="container clearfix">
            <div className="float-sm-left">
              <Link to="/support">Support</Link>
              <h1 title={ticket.id}>{ticket.summary}</h1>
              {renderTicketCreationInfo(ticket)}
            </div>
          </div>
        </header>
        <div className="container">
          <div className="row">
            <div className="col-lg-7 col-md-12">
              {/* TODO: make sure this logic makes sense. */}
              {!ticket.description ? null : (
                <section>
                  <TicketReply reply={ticket} createdField="opened" />
                </section>
              )}
              {Object.values(replies).map(reply => (
                <section key={reply.id}>
                  <TicketReply reply={reply} createdField="created" />
                </section>
              ))}
              {ticket.status === 'closed' ?
                this.renderTicketClosed() :
                this.renderTicketResponseForm()}
            </div>
            <div className="col-lg-5 col-md-12">
              <TicketHelper />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

TicketPage.propTypes = {
  dispatch: PropTypes.func.isRequired,
  ticket: PropTypes.object.isRequired,
  replies: PropTypes.object.isRequired,
};

function select(state, props) {
  const { ticketId } = props.params;
  const ticket = state.api.tickets.tickets[ticketId];
  return {
    ticket,
    replies: ticket && ticket._replies.replies,
  };
}

export default connect(select)(TicketPage);
