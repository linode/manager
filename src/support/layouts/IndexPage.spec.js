import React from 'react';
import sinon from 'sinon';
import { mount, shallow } from 'enzyme';
import { StaticRouter } from 'react-router-dom';

import { IndexPage } from '~/support/layouts/IndexPage';
import { api } from '~/data';

describe('support/layouts/IndexPage', () => {
  const sandbox = sinon.sandbox.create();
  const dispatch = sandbox.spy();

  afterEach(() => {
    sandbox.restore();
  });

  it('should render without error', () => {
    const mockDispatch = jest.fn();
    const wrapper = shallow(
      <StaticRouter>
        <IndexPage
          dispatch={mockDispatch}
          tickets={api.tickets}
        />
      </StaticRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a list of tickets', () => {
    const page = mount(
      <StaticRouter>
        <IndexPage
          dispatch={dispatch}
          tickets={api.tickets}
        />
      </StaticRouter>
    );

    const tickets = page.find('.TableRow');
    expect(tickets.length).toBe(Object.keys(api.tickets.tickets).length);
    const ticket = tickets.at(0);

    // Get id without hashtag
    const id = ticket.find('td').at(1).text()
      .substring('Ticket #'.length);
    const renderedTicket = api.tickets.tickets[id];
    const label = ticket.find('Link').at(0);
    expect(label.props().to).toBe(`/support/${id}`);
    expect(label.text()).toBe(renderedTicket.summary);

    expect(ticket.find(`#opened-by-${id}`).text()).toBe(renderedTicket.opened_by);
    expect(ticket.find(`#opened-${id} TimeDisplay`).props().time).toBe(renderedTicket.opened);
    expect(ticket.find(`#regarding-${id}`).text()).toBe(renderedTicket.entity.label);

    expect(ticket.find(`#updated-by-${id}`).text()).toBe(renderedTicket.updated_by);
    expect(ticket.find(`#updated-${id} TimeDisplay`).props().time).toBe(renderedTicket.updated);
  });
});
