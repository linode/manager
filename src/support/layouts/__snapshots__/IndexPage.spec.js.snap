// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`support/layouts/IndexPage should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StaticRouter
    basename=""
    location="/"
  >
    <IndexPage
      dispatch={[Function]}
      tickets={
        Object {
          "ids": Array [
            1,
          ],
          "pagesFetched": Array [
            1,
          ],
          "plural": "tickets",
          "singular": "ticket",
          "tickets": Object {
            "1": Object {
              "_replies": Object {
                "replies": Array [
                  Object {
                    "created": "2017-03-30T16:48:10",
                    "created_by": "support",
                    "description": "This is the first response.",
                    "id": 1,
                  },
                ],
              },
              "closed": null,
              "description": "Hullo!",
              "entity": Object {
                "label": "awkward-aimless-alpha-22",
                "type": "linode",
              },
              "id": 1,
              "opened": "2017-03-03T01:22:32",
              "opened_by": "tdude",
              "status": "closed",
              "summary": "This is my ticket!",
              "updated": "2017-03-03T01:22:32",
              "updated_by": "tdude",
            },
          },
          "totalPages": 1,
          "totalResults": 1,
        }
      }
    />
  </StaticRouter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <IndexPage
        dispatch={[Function]}
        tickets={
          Object {
            "ids": Array [
              1,
            ],
            "pagesFetched": Array [
              1,
            ],
            "plural": "tickets",
            "singular": "ticket",
            "tickets": Object {
              "1": Object {
                "_replies": Object {
                  "replies": Array [
                    Object {
                      "created": "2017-03-30T16:48:10",
                      "created_by": "support",
                      "description": "This is the first response.",
                      "id": 1,
                    },
                  ],
                },
                "closed": null,
                "description": "Hullo!",
                "entity": Object {
                  "label": "awkward-aimless-alpha-22",
                  "type": "linode",
                },
                "id": 1,
                "opened": "2017-03-03T01:22:32",
                "opened_by": "tdude",
                "status": "closed",
                "summary": "This is my ticket!",
                "updated": "2017-03-03T01:22:32",
                "updated_by": "tdude",
              },
            },
            "totalPages": 1,
            "totalResults": 1,
          }
        }
      />,
      "history": Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "tickets": Object {
          "ids": Array [
            1,
          ],
          "pagesFetched": Array [
            1,
          ],
          "plural": "tickets",
          "singular": "ticket",
          "tickets": Object {
            "1": Object {
              "_replies": Object {
                "replies": Array [
                  Object {
                    "created": "2017-03-30T16:48:10",
                    "created_by": "support",
                    "description": "This is the first response.",
                    "id": 1,
                  },
                ],
              },
              "closed": null,
              "description": "Hullo!",
              "entity": Object {
                "label": "awkward-aimless-alpha-22",
                "type": "linode",
              },
              "id": 1,
              "opened": "2017-03-03T01:22:32",
              "opened_by": "tdude",
              "status": "closed",
              "summary": "This is my ticket!",
              "updated": "2017-03-03T01:22:32",
              "updated_by": "tdude",
            },
          },
          "totalPages": 1,
          "totalResults": 1,
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <IndexPage
          dispatch={[Function]}
          tickets={
            Object {
              "ids": Array [
                1,
              ],
              "pagesFetched": Array [
                1,
              ],
              "plural": "tickets",
              "singular": "ticket",
              "tickets": Object {
                "1": Object {
                  "_replies": Object {
                    "replies": Array [
                      Object {
                        "created": "2017-03-30T16:48:10",
                        "created_by": "support",
                        "description": "This is the first response.",
                        "id": 1,
                      },
                    ],
                  },
                  "closed": null,
                  "description": "Hullo!",
                  "entity": Object {
                    "label": "awkward-aimless-alpha-22",
                    "type": "linode",
                  },
                  "id": 1,
                  "opened": "2017-03-03T01:22:32",
                  "opened_by": "tdude",
                  "status": "closed",
                  "summary": "This is my ticket!",
                  "updated": "2017-03-03T01:22:32",
                  "updated_by": "tdude",
                },
              },
              "totalPages": 1,
              "totalResults": 1,
            }
          }
        />,
        "history": Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "dispatch": [Function],
          "tickets": Object {
            "ids": Array [
              1,
            ],
            "pagesFetched": Array [
              1,
            ],
            "plural": "tickets",
            "singular": "ticket",
            "tickets": Object {
              "1": Object {
                "_replies": Object {
                  "replies": Array [
                    Object {
                      "created": "2017-03-30T16:48:10",
                      "created_by": "support",
                      "description": "This is the first response.",
                      "id": 1,
                    },
                  ],
                },
                "closed": null,
                "description": "Hullo!",
                "entity": Object {
                  "label": "awkward-aimless-alpha-22",
                  "type": "linode",
                },
                "id": 1,
                "opened": "2017-03-03T01:22:32",
                "opened_by": "tdude",
                "status": "closed",
                "summary": "This is my ticket!",
                "updated": "2017-03-03T01:22:32",
                "updated_by": "tdude",
              },
            },
            "totalPages": 1,
            "totalResults": 1,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
